src/main/java/seedu/taskell/logic/commands/UndoCommand.java

```diff
@@ -54,7 +65,35 @@ public UndoCommand(int index) {
         
         return list;
     }
+
+    //removes commandHistory with tasks not present in system
+    private void updateCommandList() {
+        if (model == null) {
+            logger.severe("model is null");
+        }
+        for (CommandHistory commandHistory: commandHistoryList) {
+            if (isCommandTypeAddOrEdit(commandHistory) 
+                    && !model.isTaskPresent(commandHistory.getTask())) {
+                commandHistoryList.remove(commandHistory);
+            } else if (isUndoEditCommand(commandHistory) 
+                    && !model.isTaskPresent(commandHistory.getOldTask())) {
+                commandHistoryList.remove(commandHistory);
+            }
+        }
+        
+    }
+
+    private boolean isCommandTypeAddOrEdit(CommandHistory commandHistory) {
+        return (commandHistory.getCommandType().contains(AddCommand.COMMAND_WORD) 
+                || commandHistory.getCommandType().contains("edit")) 
+                && !commandHistory.isRedoTrue();
```

Comment:
Magic string. Please use constant in EditCommand class
![image](https://cloud.githubusercontent.com/assets/13817632/19710246/a2c2e9e6-9b5f-11e6-915b-b3c203bdccd3.png)